File: desktop-ui/src/service/api/application.ts
import { invokeDesktop } from '@/service/tauri/client';

export interface AppInfo {
  version: string;
  platform: string;
}

export type UpdateResult =
  | { kind: 'native' }
  | {
      kind: 'release';
      version: string;
      available: boolean;
      hasApk: boolean;
      releaseUrl: string;
    };

export const getAppInfo = () => invokeDesktop<AppInfo>('get_app_info');
export const checkAppUpdate = () => invokeDesktop<UpdateResult>('check_app_update');

File: crates/qq-farm-desktop/src/commands/application.rs
//! Application metadata and the mobile release check (desktop keeps its signed updater).

use serde::Serialize;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AppInfo {
    version: String,
    platform: &'static str,
}

#[tauri::command]
pub fn get_app_info(app: tauri::AppHandle) -> AppInfo {
    AppInfo { version: app.package_info().version.to_string(), platform: std::env::consts::OS }
}

#[derive(Serialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum UpdateResult {
    Native,
    #[allow(dead_code)]
    #[serde(rename_all = "camelCase")]
    Release {
        version: String,
        available: bool,
        has_apk: bool,
        release_url: String,
    },
}

#[tauri::command]
pub async fn check_app_update(app: tauri::AppHandle) -> Result<UpdateResult, String> {
    #[cfg(desktop)]
    {
        crate::updater::check_manually(app).await;
        Ok(UpdateResult::Native)
    }
    #[cfg(mobile)]
    {
        crate::release::check(&app.package_info().version.to_string()).await
    }
}
